daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/Preprocessor.scala

```diff
@@ -51,127 +51,96 @@ private[engine] final class Preprocessor(
-              // We assume that getDependencies is always given serializable types
-              ResultError(
-                Error.Preprocessing
-                  .Internal(
-                    NameOf.qualifiedNameOfCurrentFunc,
-                    s"unserializable type ${typ.pretty}",
-                    None,
-                  )
-              )
-          }
-        case Nil =>
-          tmplToProcess0 match {
-            case tmplId :: tmplsToProcess if tmplsAlreadySeen0(tmplId) =>
-              go(Nil, tmplsToProcess, tyConAlreadySeen0, tmplsAlreadySeen0)
-            case tmplId :: tmplsToProcess =>
-              interface.lookupTemplateOrInterface(tmplId) match {
-                case Right(Left(template)) =>
-                  val typs0 = template.choices.map(_._2.argBinder._2).toList
-                  val typs1 =
-                    if (tyConAlreadySeen0(tmplId)) typs0 else Ast.TTyCon(tmplId) :: typs0
-                  val typs2 = template.key.fold(typs1)(_.typ :: typs1)
-                  go(typs2, tmplsToProcess, tyConAlreadySeen0, tmplsAlreadySeen0)
-                case Right(Right(interface)) =>
-                  val typs0 = (interface.fixedChoices.values.map(_.argBinder._2)
-                    ++ interface.methods.values.map(_.returnType)).toList
-                  val typs1 =
-                    if (tyConAlreadySeen0(tmplId)) typs0 else Ast.TTyCon(tmplId) :: typs0
-                  go(typs1, tmplsToProcess, tyConAlreadySeen0, tmplsAlreadySeen0)
-                case Left(LookupError.MissingPackage(pkgId, context)) =>
-                  pullPackage(pkgId, context)
-                case Left(error) =>
-                  ResultError(Error.Preprocessing.Lookup(error))
-              }
-            case Nil =>
-              ResultDone(tyConAlreadySeen0 -> tmplsAlreadySeen0)
-          }
-      }
+      case Nil =>
+        ResultDone.Unit
     }
 
-    go(typesToProcess0, tmplToProcess0, tyConAlreadySeen0, tmplAlreadySeen0)
-  }
+  private[this] def pullTypePackages(typ: Ast.Type): Result[Unit] =
+    collectNewPackagesFromTypes(List(typ)).flatMap(pullPackages)
+
+  private[this] def pullTemplatePackage(tyCons: List[Ref.TypeConName]): Result[Unit] =
+    collectNewPackagesFromTemplateIds(tyCons).flatMap(pullPackages)
 
   /** Translates the LF value `v0` of type `ty0` to a speedy value.
     * Fails if the nesting is too deep or if v0 does not match the type `ty0`.
     * Assumes ty0 is a well-formed serializable typ.
     */
   def translateValue(ty0: Ast.Type, v0: Value): Result[SValue] =
-    safelyRun(getDependencies(List(ty0), List.empty)) {
+    safelyRun(pullTypePackages(ty0)) {
       commandPreprocessor.valueTranslator.unsafeTranslateValue(ty0, v0)
     }
 
   private[engine] def preprocessApiCommand(
       cmd: command.ApiCommand
   ): Result[speedy.Command] =
-    safelyRun(getDependencies(List.empty, List(cmd.templateId))) {
+    safelyRun(pullTemplatePackage(List(cmd.templateId))) {
```

Comment:
So my understanding is that the essence of this change is to split the old code `getDependencies`
into two parts `pullTypePackage` and `pullTemplatePackage`
which pull packages based on references to either template of type refs.

Does this this avoid some work, or is it simply a cleaner design?

Reply:
We avoid indeed some work. 
Before we were pulling the definitions of the type constructors and the templates we encountered and we were processing recursively all the types appearing in those definition (previous lines 93 to 111 and from 135 to 147). Now we avoid completely this.